lib/nerves_hub/http_fwup_stream.ex

```diff
@@ -0,0 +1,157 @@
+defmodule NervesHub.HTTPFwupStream do
+  use GenServer
+
+  require Logger
+  alias NervesHub.Fwup
+
+  @redirect_status_codes [301, 302, 303, 307, 308]
+
+  def start_link(cb) do
+    start_httpc()
+    GenServer.start_link(__MODULE__, [cb])
+  end
+
+  def stop(pid) do
+    GenServer.stop(pid)
+  end
+
+  def get(pid, url) do
+    GenServer.call(pid, {:get, url}, :infinity)
+  end
+
+  def init([cb]) do
+    {:ok, fwup} = Fwup.start_link(self())
+    Logger.configure(level: :error)
+
+    {:ok,
+     %{
+       url: nil,
+       content_length: 0,
+       buffer: "",
+       buffer_size: 0,
+       filename: "",
+       caller: nil,
+       number_of_redirects: 0,
+       fwup: fwup,
+       callback: cb
+     }}
+  end
+
+  def handle_call({:get, _url}, _from, %{number_of_redirects: n} = s) when n > 5 do
+    GenServer.reply(s.caller, {:error, :too_many_redirects})
+    {:noreply, %{s | url: nil, number_of_redirects: 0, caller: nil}}
+  end
+
+  def handle_call({:get, url}, from, s) do
+    headers = [
+      {'Content-Type', 'application/octet-stream'}
+    ]
+
+    http_opts = [timeout: :infinity, autoredirect: false]
+    opts = [stream: :self, receiver: self(), sync: false]
+
+    :httpc.request(
```

Comment:
If were going to be adding hackney, then it doesn't make sense to continue to use :httpc here

Reply:
I guess we could tackle that in a follow up.